lib/block_view_stake.go

```diff
@@ -0,0 +1,2600 @@
+		metadata.StakeAmountNanos.IsZero() ||
+		!metadata.StakeAmountNanos.IsUint64() {
+		return errors.Wrapf(RuleErrorInvalidStakeAmountNanos, "UtxoView.IsValidStakeMetadata: ")
+	}
+	transactorDeSoBalanceNanos, err := bav.GetSpendableDeSoBalanceNanosForPublicKey(transactorPkBytes, blockHeight-1)
+	if err != nil {
+		return errors.Wrapf(err, "UtxoView.IsValidStakeMetadata: ")
+	}
+	if uint256.NewInt().SetUint64(transactorDeSoBalanceNanos).Cmp(metadata.StakeAmountNanos) < 0 {
+		return errors.Wrapf(RuleErrorInvalidStakeInsufficientBalance, "UtxoView.IsValidStakeMetadata: ")
+	}
+
+	return nil
+}
+
+func (bav *UtxoView) IsValidUnstakeMetadata(transactorPkBytes []byte, metadata *UnstakeMetadata) error {
+	// Validate TransactorPublicKey.
+	transactorPKIDEntry := bav.GetPKIDForPublicKey(transactorPkBytes)
+	if transactorPKIDEntry == nil || transactorPKIDEntry.isDeleted {
+		return errors.Wrapf(RuleErrorInvalidStakerPKID, "UtxoView.IsValidUnstakeMetadata: ")
+	}
+
+	// Validate ValidatorPublicKey.
+	validatorEntry, err := bav.GetValidatorByPublicKey(metadata.ValidatorPublicKey)
+	if err != nil {
+		return errors.Wrapf(err, "IsValidUnstakeMetadata: ")
+	}
+	if validatorEntry == nil || validatorEntry.isDeleted {
+		return errors.Wrapf(RuleErrorInvalidValidatorPKID, "UtxoView.IsValidUnstakeMetadata: ")
+	}
+
+	// Validate StakeEntry exists.
+	stakeEntry, err := bav.GetStakeEntry(validatorEntry.ValidatorPKID, transactorPKIDEntry.PKID)
+	if err != nil {
+		return errors.Wrapf(err, "UtxoView.IsValidUnstakeMetadata: ")
+	}
+	if stakeEntry == nil || stakeEntry.isDeleted {
+		return errors.Wrapf(RuleErrorInvalidUnstakeNoStakeFound, "UtxoView.IsValidUnstakeMetadata: ")
+	}
+
+	// Validate 0 < UnstakeAmountNanos <= StakeEntry.StakeAmountNanos.
+	if metadata.UnstakeAmountNanos == nil || metadata.UnstakeAmountNanos.IsZero() {
+		return errors.Wrapf(RuleErrorInvalidUnstakeAmountNanos, "UtxoView.IsValidUnstakeMetadata: ")
+	}
+	if stakeEntry.StakeAmountNanos.Cmp(metadata.UnstakeAmountNanos) < 0 {
+		return errors.Wrapf(RuleErrorInvalidUnstakeInsufficientStakeFound, "UtxoView.IsValidUnstakeMetadata: ")
+	}
+
+	return nil
+}
+
+func (bav *UtxoView) IsValidUnlockStakeMetadata(transactorPkBytes []byte, metadata *UnlockStakeMetadata) error {
+	// Validate TransactorPublicKey.
+	transactorPKIDEntry := bav.GetPKIDForPublicKey(transactorPkBytes)
+	if transactorPKIDEntry == nil || transactorPKIDEntry.isDeleted {
+		return errors.Wrapf(RuleErrorInvalidStakerPKID, "UtxoView.IsValidUnlockStakeMetadata: ")
+	}
+
+	// Validate ValidatorPublicKey.
+	validatorEntry, err := bav.GetValidatorByPublicKey(metadata.ValidatorPublicKey)
+	if err != nil {
+		return errors.Wrapf(err, "UtxoView.IsValidUnlockStakeMetadata: ")
+	}
+	if validatorEntry == nil || validatorEntry.isDeleted {
+		return errors.Wrapf(RuleErrorInvalidValidatorPKID, "UtxoView.IsValidUnlockStakeMetadata: ")
+	}
+
+	// Validate StartEpochNumber and EndEpochNumber.
+	if metadata.StartEpochNumber > metadata.EndEpochNumber {
```

Comment:
See other comments. I'd get rid of StartEpochNumber and EndEpochNumber and just automatically unlock any LockedStakeEntry that has `lockedStakeEntry.LockEpochNumber+bav.Params.StakeLockupEpochDuration > currentEpochNumber`